layer2-state-provider/history_commitment_provider.go

```diff
@@ -0,0 +1,321 @@
+// so we can ignore block challenges for this calculation.
+//
+// Let's say we want to figure out the machine start opcode index for the following inputs:
+//
+// megastep=4, kilostep=5, step=10
+//
+// We can compute the opcode index using the following algorithm for the example above.
+//
+//	  4 * (1048576)
+//	+ 5 * (1024)
+//	+ 10
+//	= 4,199,434
+//
+// This generalizes for any number of subchallenge levels into the algorithm below.
+// It works by taking the sum of (each input * product of all challenge level height constants beneath its level).
+// This means we need to start executing our machine exactly at opcode index 4,199,434.
+func (p *HistoryCommitmentProvider) computeMachineStartIndex(
+	startHeights validatedStartHeights,
+) (OpcodeIndex, error) {
+	// For the block challenge level, the machine start opcode index is always 0.
+	if len(startHeights) == 1 {
+		return 0, nil
+	}
+	// The first position in the start heights slice is the block challenge level, which is over ranges of L2 messages
+	// and not over individual opcodes. We ignore this level and start at the next level when it comes to dealing with
+	// machines.
+	heights := startHeights[1:]
+	leafHeights := p.challengeLeafHeights[1:]
+
+	// Next, we compute the opcode index. We use big ints to make sure we do not overflow uint64
+	// as this computation depends on external user inputs.
+	opcodeIndex := new(big.Int).SetUint64(0)
+	idx := 1
+	for _, height := range heights {
+		total := new(big.Int).SetUint64(1)
+		for i := idx; i < len(leafHeights); i++ {
+			total = new(big.Int).Mul(total, new(big.Int).SetUint64(uint64(leafHeights[i])))
+		}
+		increase := new(big.Int).Mul(total, new(big.Int).SetUint64(uint64(height)))
+		opcodeIndex = new(big.Int).Add(opcodeIndex, increase)
+		idx += 1
+	}
+	if !opcodeIndex.IsUint64() {
+		return 0, fmt.Errorf("computed machine start index overflows uint64: %s", opcodeIndex.String())
+	}
+	return OpcodeIndex(opcodeIndex.Uint64()), nil
+}
+
+// Computes the the number of individual opcodes we need to step through a machine at a time.
+// Each challenge level has a different amount of ranges of opcodes, so the overall step size can be computed
+// as a multiplication of all the next challenge levels needed.
+//
+// As an example, this function helps answer questions such as: "How many individual opcodes are there in a single step of a
+// Megastep challenge?"
+func (p *HistoryCommitmentProvider) computeStepSize(challengeLevel uint64) (StepSize, error) {
+	// The last challenge level is over individual opcodes, so the step size is always 1 opcode at a time.
+	if challengeLevel+1 == p.numberOfChallengeLevels() {
+		return 1, nil
+	}
+	// Otherwise, it is the multiplication of all the challenge leaf heights at the next
+	// challenge levels.
+	levels := p.challengeLeafHeights[challengeLevel+1:]
+	total := uint64(1)
+	for _, h := range levels {
+		total *= uint64(h)
```

Comment:
Should we worry about overflow here?

Reply:
Not here, because these are constants we control. Not user input